### PR TITLE
Fix/issue 895

### PR DESCRIPTION
**Pull request by narinluangrath**
What's new:
- Broke apart the bar svg path into smaller pieces
- Fixed issue 895 for vertical bars. Will add fix for horizontal bars if this code is approved.

Other comments:
- Maybe consider breaking up the code into smaller sub-functions? 
- I'm looking for suggestions to improve my code, I know it's a bit hard to follow!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -102,5 +102,8 @@
     "webpack-dev-server": "^1.10.0",
     "webpack-stats-plugin": "^0.1.1"
   },
-  "sideEffects": false
+  "sideEffects": false,
+  "dependencies": {
+    "circle2": "^1.1.1"
```
Comment (boygirl) [inline package.json]: I would really prefer not to take this dependency. Can you get it done without?

Comment (narinluangrath) [reply package.json]: I can give it a try! Do you have some general guidelines on whether or not to take a dependency, like a best practices guide? I didn't see anything about that in the `formidable-playbook`.

Comment (ryan-roemer) [reply package.json]: Assuming we did want this dep, isn't it supposed to be in https://github.com/FormidableLabs/victory/blob/master/packages/victory-bar/package.json ?

Comment (boygirl) [reply package.json]: Yes, keeping it in the main package.json would add it to every package.

File: packages/victory-bar/src/bar.js [hunk 2]
```diff
@@ -82,82 +85,118 @@ export default class Bar extends React.Component {
...
-      intrxnsRight[1]
-
-    const start = `M ${x0 + cornerRadius.bottom}, ${y0}`;
-    const bottomLeftArc = `A ${bottomArc}, ${intrxnLeft.x}, ${intrxnLeft.y}`;
-    const topLeftArc = `A ${topArc}, ${x0 + cornerRadius.top}, ${y1}`;
-    const topLine = `L ${x1 - cornerRadius.top}, ${y1}`;
-    const topRightArc = `A ${topArc}, ${intrxnRight.x}, ${intrxnRight.y}`;
-    const bottomRightArc = `A ${bottomArc}, ${x1 - cornerRadius.bottom}, ${y0}`;
-    const end = "z";
+    // There are four ways the verticalBarPath can self-intersect
```
Comment (narinluangrath) [inline packages/victory-bar/src/bar.js]: *three ways

File: package.json (hunk 1 above)
Comment (narinluangrath) [reply package.json]: Thanks for pointing that out! I'll uninstall it from the main package.json.

**Comment by narinluangrath** [comment]
**Do Not Merge!** Just pushing to get feedback.

I am considering some changes to make `getVerticalBarPath` less huge:
- turn `hasSelfIntersectingPath` into a separate helper function
- turn each of the three cases into it's own helper function

I ready though the testing section of `CONTRIBUTING.md` but I'm still not sure the best way to test the change. Maybe add a very small bar chart to `stories/victory-bar.js`?

File: packages/victory-bar/src/bar.js [hunk 3]
```diff
@@ -53,17 +55,141 @@ export default class Bar extends React.Component {
...
+      const bottomRightCenter = new Point(x1 - cornerRadius.bottom, y0 - sign * cornerRadius.bottom);
+      const bottomRightCircle = new Circle(bottomRightCenter, cornerRadius.bottom);
+
+      // circles intersect at two points, find the correct one
+      const intrxnsRight = topRightCircle.intersection(bottomRightCircle);
+      const intrxnRight = intrxnsRight.length === 1 ? intrxnsRight[0] :
+        intrxnsRight[0].x >= intrxnsRight[1].x ? intrxnsRight[0] :
+        intrxnsRight[1];
+
+      const start = `M ${x0 + cornerRadius.bottom}, ${y0}`;
```
Comment (boygirl) [inline packages/victory-bar/src/bar.js]: Some of these don't need to be defined again for each case. It looks like `start` here is the same as the non-intersecting value. I think you can save yourself a lot of lines of code by only re-defining some of these values

File: packages/victory-bar/src/bar.js [hunk 4]
```diff
@@ -53,17 +55,141 @@ export default class Bar extends React.Component {
...
+      const bottomLeftArc = `A ${bottomArc}, ${intrxnLeft.x}, ${intrxnLeft.y}`;
+      const topLeftArc = `A ${topArc}, ${x0 + cornerRadius.top}, ${y1}`;
+      const topLine = `L ${x1 - cornerRadius.top}, ${y1}`;
+      const topRightArc = `A ${topArc}, ${intrxnRight.x}, ${intrxnRight.y}`;
+      const bottomRightArc = `A ${bottomArc}, ${x1 - cornerRadius.bottom}, ${y0}`;
+      const end = "z";
+
+      return [ start,
```
Comment (boygirl) [inline packages/victory-bar/src/bar.js]: You can just return once at the end after modifying any values you need to modify for intersections. It also wont alter the final path to leave in zero radius arcs, for example.

File: packages/victory-bar/src/bar.js [hunk 5]
```diff
@@ -53,17 +55,141 @@ export default class Bar extends React.Component {
...
+
+      // find intersection of y0 and the top left arc
+      const cxLeft = x0 + cornerRadius.top;
+      const cyLeft = y1 + sign * cornerRadius.top;
+      const intrxnLeftX = -Math.sqrt(cornerRadius.top ** 2 - (y0 - cyLeft) ** 2) + cxLeft;
+
+      // find intersection of y0 and the top right arc
+      const cxRight = x1 - cornerRadius.top;
+      const cyRight = y1 + sign * cornerRadius.top;
+      const intrxnRightX = Math.sqrt(cornerRadius.top ** 2 - (y0 - cyRight) ** 2) + cxRight;
```
Comment (boygirl) [inline packages/victory-bar/src/bar.js]: Some of these look like they can be generalized into small functions for re-use

**Comment by narinluangrath** [comment]
Makes sense to me 👍Thanks for the feedback.

File: packages/victory-bar/src/geometry-helper-methods.js [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
  * A point in the 2d plane
+ * @constructor
  * @param {number} x - x coordinate
  * @param {number} y - y coordinate
  */
-function Point(x, y) {
-  this.x = x;
-  this.y = y;
-}
+const Point = function (x, y) {
```
Comment (narinluangrath) [inline packages/victory-bar/src/geometry-helper-methods.js]: Writing `function Point(x, y)` results in ESLint error `func-style`
Writing `const Point = function (x, y)` results in ESLint error `no-invalid-this` 
Which is preferred?

**Comment by narinluangrath** [comment]
@boygirl Would I be a good idea to solve issue #1119 in this PR/branch as well? It would be kinda strange to solve issue #895, and then immediately edit the same code to solve #1119. Let me know what you think of changes, and if it looks good to you, I can apply the fix to horizontal bars and solve #1119.

File: packages/victory-bar/src/geometry-helper-methods.js (hunk 6 above)
Comment (boygirl) [reply packages/victory-bar/src/geometry-helper-methods.js]: `const Point = (x, y) => ...` should pass our linting rules. Otherwise, prefer `function Point(x, y)` and add an exception for `func-style`

**Comment by boygirl** [comment]
@narinluangrath yes, I agree those changes make sense together.

**Comment by boygirl** [comment]
@narinluangrath to clarify, yes, please do add https://github.com/FormidableLabs/victory/issues/1119 into this ticket. I also think it makes sense to move these path calculations into a separate file since they're getting so lengthy. Maybe `path-helpers`?

**Comment by narinluangrath** [comment]
@boygirl Roger that, will add #1119 solution and `path-helpers` file in next commits.

**Comment by narinluangrath** [comment]
Going to push fix after work today! Code is done, just need to pass linting rules.

**Comment by boygirl** [comment]
@narinluangrath closing this one in favor of https://github.com/FormidableLabs/victory/pull/1158
